### PR TITLE
Add more detailed data from K8s and VPP NAT44

### PR DESCRIPTION
**Pull request by jmedved**

### CODE REVIEW

File: scripts/contiv-vpp-bug-report.sh [hunk 1]
```diff
@@ -100,6 +100,12 @@ master_kubectl get services -o wide --all-namespaces > k8s-services.txt
 echo " - networkpolicy"
 # Don't show 'No resources found.' error on stderr.
 master_kubectl get networkpolicy -o wide --all-namespaces > k8s-policies.txt 2>/dev/null
+echo " - statefulsets"
+# Don't show 'No resources found.' error on stderr.
+master_kubectl get statefulset -o wide --all-namespaces > k8s-statefulsets.txt 2>/dev/null
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: Can you redirect stderr to the log file as well (and while you're at it, change the commands above to do the same)?  This way any problems collecting the data will be logged for us.

File: scripts/contiv-vpp-bug-report.sh [hunk 2]
```diff
@@ -14,11 +14,20 @@ master_kubectl() {
...
 }
 
+VPP_CMD_ERROR_LOG="vpp-cmd-errors.log"
 get_vpp_data() {
     echo " - vppctl '$1'"
+    echo "- vppctl '$1':" >> "$VPP_CMD_ERROR_LOG"
     # We need to call out /usr/bin/vppctl because /usr/local/bin/vppctl is a wrapper script that doesn't work inside the
     # container.
-    master_kubectl exec "$POD_NAME" -n kube-system -c contiv-vswitch /usr/bin/vppctl "$1" > "$2.log"
+    master_kubectl exec "$POD_NAME" -n kube-system -c contiv-vswitch /usr/bin/vppctl "$1" > "$2.log" 2>>"$VPP_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: Having a single log for all errors in this fashion means we won't be able to figure out what command generated the error.

Comment (jmedved) [reply scripts/contiv-vpp-bug-report.sh]: The name of the command is echoed into the log file just before the error message. the resulting error log looks something like this:
- vppctl 'sh int':
command terminated with exit code 141
- vppctl 'sh int addr':
- vppctl 'sh ip fib':
command terminated with exit code 141
- vppctl 'sh l2fib verbose':
- vppctl 'sh ip arp':
command terminated with exit code 141
- vppctl 'sh vxlan tunnel':
- vppctl 'sh nat44 interfaces':
command terminated with exit code 141
- vppctl 'sh nat44 static mappings':
- vppctl 'sh nat44 addresses':
- vppctl 'sh nat44 deterministic mappings':
command terminated with exit code 141
- vppctl 'sh nat44 deterministic sessions':
command terminated with exit code 141
- vppctl 'sh nat44 deterministic timeouts':
command terminated with exit code 141
- vppctl 'sh nat44 hash tables detail':
command terminated with exit code 141
- vppctl 'sh nat44 sessions detail':
command terminated with exit code 141
- vppctl 'sh acl-plugin acl':
- vppctl 'sh hardware-interfaces':
command terminated with exit code 141
- vppctl 'sh errors':
command terminated with exit code 141

IMO, it's better to have all errors in one file, as long as you can identify them.

File: scripts/contiv-vpp-bug-report.sh [hunk 3]
```diff
@@ -14,11 +14,27 @@ master_kubectl() {
     ssh "$SSH_USER@$MASTER" "${SSH_OPTS[@]}" kubectl "$@" || true
 }
 
+VPP_CMD_ERROR_LOG="vpp-cmd-errors.log"
 get_vpp_data() {
     echo " - vppctl '$1'"
+    echo "- vppctl '$1':" >> "$VPP_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: A single echo using tee would be better than repeating the same message:
```
echo "- vppctl '$1'" | tee -a "$VPP_CMD_ERROR_LOG"
```

File: scripts/contiv-vpp-bug-report.sh (hunk 2 above)
Comment (nbartos) [reply scripts/contiv-vpp-bug-report.sh]: Nitpick: spacing after >> should be consistent with the rest of the file:
```
2>> "$VPP_CMD_ERROR_LOG"
```

File: scripts/contiv-vpp-bug-report.sh [hunk 4]
```diff
@@ -14,11 +14,27 @@ master_kubectl() {
...
     # We need to call out /usr/bin/vppctl because /usr/local/bin/vppctl is a wrapper script that doesn't work inside the
     # container.
-    master_kubectl exec "$POD_NAME" -n kube-system -c contiv-vswitch /usr/bin/vppctl "$1" > "$2.log"
+    master_kubectl exec "$POD_NAME" -n kube-system -c contiv-vswitch /usr/bin/vppctl "$1" > "$2.log" 2>>"$VPP_CMD_ERROR_LOG"
+}
+
+ETCD_CMD_ERROR_LOG="etcd-cmd-errors.log"
+get_etcd_data() {
+    echo " - etcdctl '$1'"
+    echo "- etcdctl '$1':" >> "$ETCD_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: Same tee comment as above.

File: scripts/contiv-vpp-bug-report.sh [hunk 5]
```diff
@@ -14,11 +14,27 @@ master_kubectl() {
...
+    master_kubectl exec "$POD_NAME" -n kube-system -- sh -c "$1" > "$2.log" 2>>"$ETCD_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: Same spacing comment as above.

File: scripts/contiv-vpp-bug-report.sh [hunk 6]
```diff
@@ -14,11 +14,27 @@ master_kubectl() {
...
+}
+
+K8S_CMD_ERROR_LOG="k8s-cmd-errors.log"
+get_k8s_data() {
+    echo " - $1"
+    echo "- $1:" >> "$K8S_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: tee

File: scripts/contiv-vpp-bug-report.sh [hunk 7]
```diff
@@ -14,11 +14,27 @@ master_kubectl() {
...
+    master_kubectl "$2" > "$3" 2>>"$K8S_CMD_ERROR_LOG"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: spacing

File: scripts/contiv-vpp-bug-report.sh [hunk 8]
```diff
@@ -133,9 +152,30 @@ for POD in $PODS; do
     # get_vpp_data "api trace custom-dump /tmp/trace.api" api-trace-dump
     echo
     popd >/dev/null
-
 done
 
+KSR_POD="$(master_kubectl get po -n kube-system -l k8s-app=contiv-ksr -o "'go-template={{range .items}}{{printf \"%s,%s\" (index .metadata).name (index .spec).nodeName}}{{end}}'")"
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: Can you please break up lines over 120 chars?  Super long ones are hard to read.

File: scripts/contiv-vpp-bug-report.sh [hunk 9]
```diff
@@ -133,9 +152,30 @@ for POD in $PODS; do
...
+popd >/dev/null
+
+CONTIV_ETCD_POD="$(master_kubectl get po -n kube-system -l k8s-app=contiv-etcd -o "'go-template={{range .items}}{{printf \"%s,%s\" (index .metadata).name (index .spec).nodeName}}{{end}}'")"
+# Get contiv-etcd logs
+IFS=',' read -r POD_NAME NODE_NAME <<< "$CONTIV_ETCD_POD"
+echo "Collecting Kubernetes data for pod $POD_NAME on node $NODE_NAME:"
+mkdir -p "$NODE_NAME"
+pushd "$NODE_NAME" >/dev/null
+# Get contiv-etcd data
+get_etcd_data "\"export ETCDCTL_API=3; etcdctl --endpoints=127.0.0.1:32379 get \"/\" --prefix=true\"" etcd_tree
```
Comment (nbartos) [inline scripts/contiv-vpp-bug-report.sh]: You should be able to use a combination of ' and " to make this easier to read. Also, you should only use ; when you don't want the failure code to propagate.  Something like:
```
get_etcd_data '"export ETCDCTL_API=3 && etcdctl --endpoints=127.0.0.1:32379 get "/" --prefix=true"' etcd_tree
```

**Pull request merged**